### ISSUE TITLE
Long-lived KnownHostsFile instance will clobber changes made by others if its `save` method is used

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6256]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#6256|
|Type|defect|
|Created|2013-01-13 22:53:22Z|
|Branch|https://github.com/twisted/twisted/tree/knownhostsfile-6256|

Entries from the file are only read in `KnownHostsFile.fromPath` - ie, at construction time.  Any future changes made to the file, either by other code in the same process or (perhaps more likely) other processes, will be wiped out as soon as `save` is used.  In particular, consider a long-running `KnownHostsFile`-using process which shares a known_hosts file with a user with their own command line `ssh` process.  Any changes (eg, keys added to the file by `ssh` or deleted by `ssh-keygen -R`) will be lost each time `KnownHostsFile.save` is invoked.

Instead, some minimal level of safety could be introduced by re-reading the file at the beginning of `save` and merging the on-disk data into the in-memory data (which can only be additions given the current `KnownHostsFile` API).

An optimization to this logic might be to check the mtime of the file at strategic points and use this as an indicator that the file may have changed and needs to be re-read.

An internal refactoring that will probably make the logic for this simpler would be to keep additions separate from the list of entries read from (or most recently written to) the underlying file.  This makes the merge much easier (throw away the list of persisted entries and replace it with the contents of the file at the beginning of save) and also points the way towards supporting _removals_, which might be a nice API addition.

<details><summary>Searchable metadata</summary>

```
trac-id__6256 6256
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_knownhostsfile_6256 branches/knownhostsfile-6256
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1358117602000000 1358117602000000
changetime__1362096523000000 1362096523000000
version__None None
owner__exarkun exarkun
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [37026]) Branching to 'knownhostsfile-6256'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

(In [37033]) Restore the behavior of KnownHostsFile(path)

I think.

This behavior is kind of silly.  I do not really want to provide it,
particularly not by default.  But... backwards compatibility.  Perhaps
it should at least be deprecated.


refs [#6256](https://github.com/twisted/twisted/issues/6256)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/knownhostsfile-6256)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [37038]) Satisfy Python 3.x

There is no cStringIO on 3.x.  `io.BytesIO` is the 2.6, 2.7, 3.3 compatible spelling of
the name of the object that's desired here.  Necessary since twisted.names.dns tests
depend on this module now, and twisted.names.dns is supposed to work on Python 3.x

refs [#6256](https://github.com/twisted/twisted/issues/6256)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_6"></a>|@tomprince set owner to @exarkun|
|-|-|

  1, The implementation has the strange behavior that ignores external changes until `.save()` is 
called, then starts reflecting them. I think this is liable to cause surprise.
  1. `test_defaultInitializerClobbersExisting` seems to test that the underlying file is actually clobbered only indirectly. Particularly in view of 1, it would be better to test that the underlying file is as we expect it.
  1. (minor) `HashedEntryTests.test_equality` is getting run twice, since `HashedEntryWithCommentTests` inherits from `HashedEntryTests`.
  1. There isn't a test of an added host with mismatched host-key.
Please resubmit for review, after addressing the above issues.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun removed owner|
|-|-|

Thanks.

> 1, The implementation has the strange behavior that ignores external changes until .save() is called, then starts reflecting them. I think this is liable to cause surprise.

If `KnownHostsFile.fromPath` is used to construct the instance, then this isn't the case.  The contents of the file will be reflected by `iterentries` (therefore also `hasHostKey` and `verifyHostKey`).  If `KnownHostsFile.__init__` is used directly, then the contents will indeed be ignored, but they'll also be overwritten by the first call to `save`.  This makes me think it is correct to ignore them.

I don't particularly like this API or behavior, but it's basically the same as the current trunk behavior without these changes, so I think retaining them makes sense (though perhaps replacing these `KnownHostsFile` APIs with something new and less surprising would also make sense).

> test_defaultInitializerClobbersExisting seems to test that the underlying file is actually clobbered only indirectly. Particularly in view of 1, it would be better to test that the underlying file is as we expect it.

Fixed in r37319.

> (minor) HashedEntryTests.test_equality is getting run twice, since HashedEntryWithCommentTests inherits from HashedEntryTests.

True.  This is a good example demonstrating how having subclasses of TestCases is not ideal.  I'd rather not refactor these classes as part of this branch.  I can file a ticket for this issue.

> There isn't a test of an added host with mismatched host-key.

I'm not sure what this means.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_8"></a>|@tomprince set owner to @exarkun|
|-|-|

>> 1, The implementation has the strange behavior that ignores external changes until .save() is called, then starts reflecting them. I think this is liable to cause surprise.
> If KnownHostsFile.fromPath is used to construct the instance, then this isn't the case. The contents of the file will be reflected by iterentries (therefore also hasHostKey and verifyHostKey). If KnownHostsFile.__init__ is used directly, then the contents will indeed be ignored, but they'll also be overwritten by the first call to save. This makes me think it is correct to ignore them.
The unexpected behavior I am commenting on is demonstrated by the following script.
```
!python
from twisted.conch.client.knownhosts import KnownHostsFile
from twisted.conch.ssh.keys import Key
from twisted.python.filepath import FilePath

keyString = "ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCVU1feuPKpCnAR2WuVmTKUM5l7rEvansUu6WdypCKcM05EwaWGih6v0hHx+PzeJ5vjt2RfVXei3o664hYXKCYsQ0a3ROOGnF0kRu3gZKqMQa9ubndCUQUeLqGzV6aYZigjPs/B0Nd/MuyV34+9pjSTWSNXV1D6abK/e+bDY+l+RUCnJPpGfr4Rbys7XiafjAdU83l++kVhC9vOzlGBU5pDZ1dCTrSDw86YzPuThtxajy2xyjpdy4VXg1iLKuKqerTjzBTj2VsQJyKnZzeOeYkUffcBjfxmy/0ACq3FeKg2cr7GTb9Q7pQSSduh/uTTamhrQGTyYDsQhpCjih2468I9 example"
key = Key.fromString(keyString)
khf = KnownHostsFile(FilePath("known_hosts"))
khf.addHostKey("example.net", key)
FilePath("known_hosts").setContent("example.com " + keyString)
print khf.hasHostKey("example.com", key) # False
print khf.hasHostKey("example.net", key) # True
khf.save()
FilePath("known_hosts").setContent("example.com " + keyString)
print khf.hasHostKey("example.com", key) # True but expect False
print khf.hasHostKey("example.net", key) # False but expect True
```
To get the expected behavior (which I agree isn't sensible), I think you just need to not reset `_clobber` or `_added` if `_clobber == True`.

>>  There isn't a test of an added host with mismatched host-key.
> I'm not sure what this means.
```
!python
#psudeo-code
f = FilePath(...)
khf = KnownHostFile.fromPath(f)
f.setContent("example.com ssh-rsa AAAAA")
knf.hasHostKey("example.com", key.fromString("ssh-rsa BBBBB")) # Raises
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

Thanks.

> The unexpected behavior I am commenting on is demonstrated by the following script.

I guess I would like to call that behavior a bug as well and change it.  The trunk behavior without this change is to make changes to the underlying file invisible (and, by being unaware of them, clobber them sometimes).

The behavior of the branch is better, it reflects changes to the contents of the file and avoids wiping those changes out when `save` is called - except for the documented case where `__init__` is used directly to explicitly create an "empty" `KnownHostsFile`.

> There isn't a test of an added host with mismatched host-key.

There is a test for a mismatched host key, though.  I don't think the source of the host key is terribly important, since `hasHostKey` is implemented with `iterentries` and `iterentries` is tested to verify that it returns `PlainEntry` or `HashedEntry` instances either from the file or from the in-memory `_added` buffer.

But:

  1. It's an easy test to add
  1. It's sort of nonsensical for the `HostKeyMismatch` exception to have lineno and path information in the case of in-memory entries
  1. There was a bug in the line number calculation for saved entries when any unsaved entries existed

So I added a couple tests and fixed that bug in r37326.  I'm not exactly thrilled by the fix for the bug, but it's a small change and this approach avoids having to make `iterentries` more complicated or introduce another helper for `hasHostKey` to use.  I'm open to suggestions for other approaches.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_10"></a>|@tomprince set owner to @exarkun|
|-|-|

> - except for the documented case where __init__ is used directly to explicitly create an "empty" KnownHostsFile.
That is the behavior that my example script is exercising and which has (to me) surprising behavior. But, I don't care either way, and nobody should be using this anyway.

> I'm not exactly thrilled by the fix for the bug, but it's a small change and this approach avoids having to make iterentries more complicated or introduce another helper for hasHostKey to use. I'm open to suggestions for other approaches.
I agree with your sentiment, but don't have any better ideas. (Maybe add a comment that `lineidx` is 0 based and `line` is 1 based?

Go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [37334]) Merge knownhostsfile-6256

Author: exarkun
Reviewer: tom.prince
Fixes: [#6256](https://github.com/twisted/twisted/issues/6256)

Change `twisted.conch.client.knownhosts.KnownHostsFile` so that it preserves
changes made to the underlying file made by means other than its own
`addHostKey` API and so that it reflects those changes in future
`hasHostKey` and `verifyHostKey` calls.
